pkg/storage/bucket/delayed_bucket_client.go

```diff
@@ -0,0 +1,83 @@
+// SPDX-License-Identifier: AGPL-3.0-only
+
+package bucket
+
+import (
+	"context"
+	"io"
+	"math/rand"
+	"time"
+
+	"github.com/thanos-io/thanos/pkg/objstore"
+)
+
+// DelayedBucketClient wraps objstore.InstrumentedBucket and add a random delay to each API call.
+// This client is intended to be used only for testing purposes.
+type DelayedBucketClient struct {
+	wrapped  objstore.Bucket
+	minDelay time.Duration
+	maxDelay time.Duration
+}
+
+func NewDelayedBucketClient(wrapped objstore.Bucket, minDelay, maxDelay time.Duration) objstore.Bucket {
+	if minDelay < 0 || maxDelay < 0 || maxDelay < minDelay {
+		// We're fine just panicking, because we expect this client to be used only for testing purposes.
+		panic("invalid delay configuration")
+	}
+
+	return &DelayedBucketClient{
+		wrapped:  wrapped,
+		minDelay: minDelay,
+		maxDelay: maxDelay,
+	}
+}
+
+func (m *DelayedBucketClient) Upload(ctx context.Context, name string, r io.Reader) error {
+	m.delay()
+	return m.wrapped.Upload(ctx, name, r)
+}
+
+func (m *DelayedBucketClient) Delete(ctx context.Context, name string) error {
+	m.delay()
+	return m.wrapped.Delete(ctx, name)
+}
+
+func (m *DelayedBucketClient) Name() string {
+	return m.wrapped.Name()
+}
+
+func (m *DelayedBucketClient) Iter(ctx context.Context, dir string, f func(string) error, options ...objstore.IterOption) error {
+	m.delay()
+	return m.wrapped.Iter(ctx, dir, f, options...)
+}
+
+func (m *DelayedBucketClient) Get(ctx context.Context, name string) (io.ReadCloser, error) {
+	m.delay()
```

Comment:
The delay should be deferred too, which IMO is more important, as this way we simulate the delay of observation of the real world (API told us that there was no item, but while we got the response, one could be created).

I think the best option is to do both:
```go
m.delay()
defer m.delay()
```